banyand/internal/storage/index.go

```diff
@@ -224,3 +229,84 @@ func appendSeriesList(dest, src pbv1.SeriesList, filter posting.List) pbv1.Serie
 func (s *seriesIndex) Close() error {
 	return s.store.Close()
 }
+
+type seriesIndexController[T TSTable, O any] struct {
+	ctx       context.Context
+	hot       *seriesIndex
+	standby   *seriesIndex
+	scheduler *timestamp.Scheduler
+	l         *logger.Logger
+	opts      TSDBOpts[T, O]
+}
+
+func newSeriesIndexController[T TSTable, O any](
+	ctx context.Context,
+	opts TSDBOpts[T, O],
+) (*seriesIndexController[T, O], error) {
+	location := filepath.Clean(opts.Location)
+	path := fmt.Sprintf("idx-%v", fastrand.Uint32())
+	h, err := newSeriesIndex(ctx, filepath.Join(location, path), opts.SeriesIndexFlushTimeoutSeconds)
+	if err != nil {
+		return nil, err
+	}
+	path = fmt.Sprintf("idx-%v", fastrand.Uint32())
+	sb, err := newSeriesIndex(ctx, filepath.Join(location, path), opts.SeriesIndexFlushTimeoutSeconds)
+	if err != nil {
+		return nil, err
+	}
+	l := logger.Fetch(ctx, "seriesIndexController")
+
+	var expr string
+	switch opts.TTL.Unit {
+	case HOUR:
+		expr = "5 *"
+	case DAY:
+		expr = "5 0"
+	}
+	clock, _ := timestamp.GetClock(ctx)
+	scheduler := timestamp.NewScheduler(l, clock)
+	sir := &seriesIndexController[T, O]{
+		hot:       h,
+		standby:   sb,
+		ctx:       ctx,
+		scheduler: scheduler,
+		l:         l,
+	}
+	if err := scheduler.Register("seriesIndexController", cron.Minute|cron.Hour, expr, sir.run); err != nil {
+		return nil, err
+	}
+	return sir, nil
+}
+
+func (sir *seriesIndexController[T, O]) run(_ time.Time, l *logger.Logger) bool {
+	err := sir.hot.Close()
+	if err != nil {
+		l.Error().Msg("fail to close old hot series index")
+		return false
+	}
+	sir.hot = sir.standby
+	location := filepath.Clean(sir.opts.Location)
+	path := fmt.Sprintf("idx-%v", fastrand.Uint32())
+	sir.standby, err = newSeriesIndex(sir.ctx, filepath.Join(location, path), sir.opts.SeriesIndexFlushTimeoutSeconds)
+	if err != nil {
+		l.Error().Msg("fail to create series index")
+		return false
+	}
+	return true
```

Comment:
The current execution order is incorrect. The correct order should be as follows:

1. Swap the hot and standby, and ensure that an RWLock is used to protect the swap.
2. Wait for an hour, then shut down the new standby (which was the previous hot).
3. If the current time is less than an hour away from the next TTL, create a new standby.